### PR TITLE
Add psalm assertions

### PR DESCRIPTION
**Pull request by muglug**
These allow anyone using Psalm (or any other tool that understands its annotations) to understand the effect of `beberlei/assert` calls in their code.

Similar to https://github.com/sebastianbergmann/phpunit/pull/3708

### CODE REVIEW

File: lib/Assert/Assertion.php [hunk 1]
```diff
@@ -1834,6 +1920,8 @@ public static function e164($value, $message = null, $propertyPath = null)
      * @param string|null                                        $message
      * @param string|null                                        $propertyPath
      *
+     * @psalm-assert non-empty-countable $value
```
Comment (simPod) [inline lib/Assert/Assertion.php]: This gives me

```
PHP Fatal error:  Uncaught Psalm\Exception\TypeParseTreeException: no hyphens allowed in projectdir/vendor/vimeo/psalm/src/Psalm/Type/Atomic.php:187
Stack trace:
#0 projectdir/vendor/vimeo/psalm/src/Psalm/Type.php(129): Psalm\Type\Atomic::create('Assert\\non-empt...', NULL, Array)
#1 projectdir/vendor/vimeo/psalm/src/Psalm/Type.php(102): Psalm\Type::parseTokens(Array, NULL, Array)
#2 projectdir/vendor/vimeo/psalm/src/Psalm/Type/Reconciler.php(1123): Psalm\Type::parseString('Assert\\non-empt...', NULL, Array)
#3 projectdir/vendor/vimeo/psalm/src/Psalm/Type/Reconciler.php(205): Psalm\Type\Reconciler::reconcileTypes('Assert\\non-empt...', Object(Psalm\Type\Union), '$value', Object(Psalm\Internal\Analyzer\StatementsAnalyzer), false, Array, Object(Psalm\CodeLocation), Array, 0)
#4 projectdir/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/Statements/Expression/CallAnalyzer. in projectdir/vendor/vimeo/psalm/src/Psalm/Type/Atomic.php on line 187
```

Comment (muglug) [reply lib/Assert/Assertion.php]: Sorry - fixed in 7a14e87 - will release tonight

Comment (simPod) [reply lib/Assert/Assertion.php]: Don't be sorry ;) Thanks

**Comment by rquadling** [comment]
What's the benefit these annotations? Who will maintain them when we add new assertions? What's the consequences of not adding them? What's the consequence/diagnostics of an incorrect one (say we edit the method signature or behaviour)?

**Comment by muglug** [comment]
> What's the benefit these annotations?

The annotations allow Psalm to understand what effect these assertions when it analyses these calls during type inference.

> What’s the consequence of not adding them?

A mild inconvenience for people who rely on this library and Psalm.

> What's the consequence/diagnostics of an incorrect one (say we edit the method signature or behaviour)?

The few annotations here are relatively simple - if you were to change the assertion operation drastically (e.g. returning false vs throwing an error) the assertions would have to be changed, but that would constitute a very breaking change for this library anyway.

**Comment by rquadling** [comment]
Psalm vs PHPStan? Do any IDEs support Psalm?

**Comment by muglug** [comment]
> Psalm vs PHPStan? Do any IDEs support Psalm?

Psalm runs a language server that is compatible with most IDEs: https://psalm.dev/docs/running_psalm/language_server/

There's also this ticket: https://youtrack.jetbrains.com/issue/WI-45248

PHPStan doesn't support any docblock assertion syntax, but it may in the future.

**Comment by emilioastarita** [comment]
:+1:  These annotations could be really useful for us. We use Assert extensively and we added psalm recently to our pipeline.

**Comment by rquadling** [comment]
Can the merge conflict be resolved (rebasing onto the latest master too please).

**Comment by orklah** [comment]
It would be cool to have those assertions. I intended to use this package to validate my input to be fully typed in psalm but it doesn't help without those.

**Comment by githoober** [comment]
Checking again today. Is there anything we can do to have this PR merged?

**Comment by githoober** [comment]
@rquadling this branch does not seem to have conflicts with the master. Could you merge it?

**Comment by magnetik** [comment]
It now have a conflict, do you mind rebasing it?

**Comment by Wirone** [comment]
@muglug maybe you can close this issue since @beberlei rebased your work and merged it in #305 :slightly_smiling_face:

**Comment by beberlei** [comment]
Merged in #305

**Pull request closed without merging**